csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,290 @@
+        continue;
+      }
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        for (const ValGroup& vg_inp : graph_->inputGroups(eg)) {
+          vg_stack.push(vg_inp);
+        }
+      }
+    }
+  }
+
+  //! Look at all ValGroups that are producers of abstract_tensor_.domain and
+  //! map any of those containing loop IterDomains in tv to those loop
+  //! IterDomains. These provide the starting points for scheduling.
+  GroupIdMap mapScheduledGroupsToLoopIterDomains(TensorView* tv) const {
+    GroupIdMap tv_ids;
+
+    // Look up each loop IterDomain in tv and assert that it is in
+    // scheduled_val_groups_, then map it.
+    for (IterDomain* loop_id : tv->getLoopDomain()) {
+      ValGroup vg = graph_->toGroup(loop_id);
+      // NOTE: this check guarantees that we will only ever _append_ transforms
+      // to the loop domain of tv. If instead we searched for producer IDs in
+      // the root->loop path of tv, then we could wind up silently erasing
+      // some pre-existing transforms.
+      NVF_CHECK(
+          scheduled_val_groups_.count(vg) > 0,
+          "Found loop IterDomain ",
+          loop_id->toString(),
+          " that is not described by abstract tensor.");
+      tv_ids.emplace(vg, loop_id);
+    }
+
+    return tv_ids;
+  }
+
+  IterDomain* replayAbstractId(AbstractId abs_id, GroupIdMap& tv_ids) {
+    ValGroup g = abstractIdToValGroup(abs_id);
+
+    // This holds ValGroups that we cannot compute from tv's root. When
+    // we detect that any inputs of an ExprGroup are in these groups, we know
+    // that we cannot compute that ExprGroup so we should skip it.
+    //
+    // For example, suppose we had
+    //   tv:
+    //     logical/loop: iS0{i0} iS1{i1} iS2{i2}
+    //
+    //   graph:
+    //     ValGroup 0: iS0
+    //     ValGroup 1: iS1
+    //     ValGroup 2: iS2 iS6
+    //     ValGroup 3: iS3
+    //     ValGroup 4: iS4
+    //     ValGroup 5: iS5
+    //     ValGroup 6: iS7
+    //     ValGroup 7: iS8
+    //     ExprGroup 0:
+    //       iS3 = merge(iS0, iS1)
+    //     ExprGroup 1:
+    //       iS4, iS5 = split(iS3, 16)
+    //     ExprGroup 2:
+    //       iS8 = merge(iS6, iS7)
+    //
+    //   abstract_tensor_.domain:
+    //     ValGroup 4, ValGroup 5, ValGroup 7
+    //
+    // In this case, ValGroups 4 and 5 are computable since those ValGroups are
+    // produced by ExprGroup 1 which itself produced by ExprGroup 0, and
+    // tv includes iS0 and iS1.
+    //
+    // However, ValGroup 7 is not computable. It is produced by ExprGroup 2
+    // whose producer ValGroups are 2 and 6. ValGroup 2 is computable since iS0
+    // is in tv, however there is no IterDomain in tv that can be
+    // used to represent ValGroup 6 which also has no producer ValGroups.
+    std::unordered_set<ValGroup> uncomputable_groups;
+
+    std::stack<ValGroup> vg_stack({g});
+    // Strategy: fill out tv_ids in the direction of g
+    auto propagate = [&](ValGroup vg) {
```

Comment:
This lambda seems quite complex. It seems it does multiple things, including finding proper replay inputs, actually replaying domains, etc. I'd suggest rewriting the lambda such that the semantic difference would be more explicitly highlighted.

Reply:
You're right. I removed the lambda and tried to streamline the body of the stack loop. Moving out `replayIdExpr` helped. I also added some more comments and changed the name of the stack to `eval_stack` since it's the stack of `ValGroup`s to try evaluating next.